### PR TITLE
Add Rolling Upgrade detection capability in LR

### PR DESCRIPTION
**Pull request by hisundar**
## Overview

Description:

Why should this be merged: 
LR needs to adhere to a mode where not all cluster nodes are upgraded to the latest (current) version.
This PR only adds the interface to detect if we are in a version pinned phase.
Handling of the version pinning will follow.

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.ILogReplicationVersionAdapter;
+import org.corfudb.runtime.collections.TxnContext;
+
+/**
+ * Rolling upgrade handling means cluster must function in a mode where not all
+ * nodes are running the same codebase.
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once upgrades are done it may atomically migrate data, clean up old code and switch
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:11]: nit: 'Once upgrades are done' -> 'Once all nodes have upgraded'?

'clean up old code' -> 'clean up old data'?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/ILogReplicationVersionAdapter.java [hunk 2]
```diff
@@ -12,4 +14,11 @@ public interface ILogReplicationVersionAdapter {
      * @return Version string that indicates the version of LR.
      */
     String getVersion();
+
+    /**
+     * @return true if the cluster is undergoing rolling upgrades with
+     * some nodes running an older code base.
+     * false if all the nodes in the cluster are at the same version as me.
+     */
+    boolean isUpgradeRunning(TxnContext txnContext);
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/ILogReplicationVersionAdapter.java:23]: would it be clearer to rename this method to something like 'isRollingUpgradeInProgress' or 'isUpgradeInProgress'?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java [hunk 3]
```diff
@@ -27,18 +30,59 @@ public String getVersion() {
...
+    /**
+     * @return false if the cluster version is already at my version
+     * and true otherwise indicating that cluster is undergoing a version upgrade
+     */
+    @Override
+    public boolean isUpgradeRunning(TxnContext txnContext) {
+        return versionString.equals(getVersion(txnContext));
+    }
+
+    public static void startRollingUpgrade(CorfuStore corfuStore) {
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:62]: Can you add an annotation or comment here that this method is for testing purposes only?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 1 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:11]: done. thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/ILogReplicationVersionAdapter.java (hunk 2 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/ILogReplicationVersionAdapter.java:23]: sure, changed.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java (hunk 3 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:62]: done. thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which is intended to be used as follows
+ * on LR startup and when upgrade is complete.
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Practically this means any code that is running Active/Standby logic is considered "V1"
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
```
Comment (xcchang) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:19]: Does this mean the workflow is this handler is mainly for the upgrade from "product with pre corfu-0.4.0.1" to "product with post corfu-0.4.0.1"? I think it is for generic rolling upgrade from "version1" to "version2" right?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            return false;
+        }
+        boolean isClusterUpgradeInProgress = versionAdapter.isUpgradeRunning(txnContext);
+        if (isClusterUpgradeInProgress) {
+            return true;
+        } // else implies cluster upgrade has completed
+        migrateData(txnContext);
+        isClusterAtV2 = true;
+        return false;
+    }
```
Comment (xcchang) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:55]: Sorry I missed the initial discussion of this change, I'm a bit confused by why the `migrateData()` is specific to LR? It looks like the data migration is something to be performed at Corfu cluster level after rolling upgrade is done? i.e. read state from old format and write in new format, delete old formats, etc.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    private volatile boolean isClusterAtV2 = false;
+    ILogReplicationVersionAdapter versionAdapter;
+    public LRRollingUpgradeHandler(ILogReplicationVersionAdapter versionAdapter) {
+        this.versionAdapter = versionAdapter;
+    }
+
+    public boolean isLRUpgradeInProgress(TxnContext txnContext) {
+        if (isClusterAtV2) {
+            return false;
+        }
```
Comment (xcchang) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:47]: So it looks like here `isClusterAtV2` indicates not only if the rolling upgrade is complete or not, it also indicates the data migration is done, and LR in this cluster could start to operate completely in V2.

So maybe we can rename it to be something more self-explanatory like `isLRReadyForV2` or `isClusterReadyForV2`? Since `isClusterAtV2` gives me an impression that it has pretty similar semantics as `isClusterUpgradeInProgress`

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 4 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:19]: yes. it is for upgrade from `corfu-0.4.0.1` to the current branch. Note that right now the current branch is master but in the future we could branch off into something like `corfu-0.4.1.0` that's what we are referring to V2 here.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 6 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:47]: Yes `isClusterAtV2` indicates both that rolling upgrade has completed & data migration is complete OR it could mean that the entire cluster is a new instantiation at V2 with no necessity for upgrade.

Sure I will rename it to `isLRReadyForV2` if that seems more self-explanatory. thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 5 above)
Comment (xcchang) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java]: Got it now, thanks!

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java [hunk 7]
```diff
@@ -27,18 +31,67 @@ public String getVersion() {
...
+        return versionString;
+    }
+
+    /**
+     * @return false if the cluster version is already at my version
+     * and true otherwise indicating that cluster is undergoing a version upgrade
+     */
+    @Override
+    public boolean isRollingUpgradeInProgress(TxnContext txnContext) {
+        return versionString.equals(getVersion(txnContext));
```
[the remark this answers is not in the document]
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:60]: thanks - good catch! this is for testing only really

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Practically this means any code that is running Active/Standby logic is considered "V1"
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
+ *                  ┌─────────────────────────────────┐
+ *                  │ isLRReadyForV2 = false (in-mem) │
+ *                  └──────────────────┬──────────────┘
+ *                                     │
+ *                     ┌───────────────▼───────────┐
+ *                   ┌─┤                           ├─┐
```
Comment (annym) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:25]: If rolling upgrade is off is not enough to say isLRReadyForV2 = true, it needs to happen that nodes are on V2, right? II think more info needs to be added.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        boolean isClusterUpgradeInProgress = versionAdapter.isRollingUpgradeInProgress(txnContext);
+        if (isClusterUpgradeInProgress) {
+            return true;
+        } // else implies cluster upgrade has completed
+        migrateData(txnContext);
+        isLRReadyForV2 = true;
+        return false;
+    }
+
+    public void migrateData(TxnContext txnContext) {
```
Comment (annym) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:57]: Missing Java Doc

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java [hunk 10]
```diff
@@ -27,18 +31,67 @@ public String getVersion() {
...
+                LogReplicationStreams.VersionString.newBuilder()
+                        .setName("VERSION").build();
+        versionString =
+                ((LogReplicationStreams.Version)txn.getRecord(VERSION_TEST_TABLE,
+                        versionStringKey).getPayload()).getVersion();
+        return versionString;
+    }
+
+    /**
+     * @return false if the cluster version is already at my version
```
Comment (annym) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:55]: Question, there is no such thing as the cluster version as the code is right now, isn't it? the version stored is the one read from a node. So we should actually take in the change Chris will be working on where we differentiate version per node and check that all are the same..

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (annym) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:60]: +1 (though I'm not sure if we have the notion of cluster version vs. node version)

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java [hunk 11]
```diff
@@ -937,4 +938,119 @@ private void verifyVersion(CorfuStore corfuStore, String expectedVersion, boolea
...
+        corfuStoreStandby.subscribeListener(standbyListener, LogReplicationMetadataManager.NAMESPACE,
+                LogReplicationMetadataManager.LR_STATUS_STREAM_TAG);
+
+        // Trigger a snapshot sync by stopping the active LR and running a CP+trim
+        stopActiveLogReplicator();
+        checkpointAndTrim(true);
+        pluginConfigFilePath = TEST_PLUGIN_CONFIG_PATH_ACTIVE;
+        startActiveLogReplicator();
+
+        // Verify that snapshot sync between the different versions was
```
Comment (annym) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:1030]: What two different versions? here both are the same version..

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java [hunk 12]
```diff
@@ -937,4 +938,119 @@ private void verifyVersion(CorfuStore corfuStore, String expectedVersion, boolea
         Assert.assertEquals(isUpgraded, actualUpgradedFlag);
         log.info("Verified version");
     }
+
+    @Test
+    public void testLocalClusterRollingUpgrade() throws Exception {
```
Comment (annym) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:943]: I don't understand what this test is validating, as all 3 nodes share the same VERSION_STRING, how are we exercising rolling upgrade?... Also, this seems a  copy/paste of another test, can't we just modify that code a bit to accept if rollingUpgrade or not, cause at the end this is just testing replication E2E when one node is upgraded and not all, is it?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once all nodes have upgraded it may atomically migrate data, clean up old code and switch
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which is intended to be used as follows
+ * on LR startup and when upgrade is complete.
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Practically this means any code that is running Active/Standby logic is considered "V1"
```
[the remark this answers is not in the document]
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:18]: I agree that seems more sound - changing the comment now. thank you

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
+ *                  ┌─────────────────────────────────┐
+ *                  │ isLRReadyForV2 = false (in-mem) │
```
[the remark this answers is not in the document]
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:21]: I think it is better if we go back to isClusterAllAtV2 with documentation indicating that this variable is set to true only when _all_ the cluster nodes are running the current or latest codebase.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 8 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:25]: got it - I am changing this to `isClusterAllAtV2` as per previous comment and previous patch that we had.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 9 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:57]: fixed thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java (hunk 10 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:55]: Yes this change depends on Chris's PR, but even then the concept of cluster version is something that will be obtained from the Version Pluging adapter. So you are right that within LR there isn't a cluster version as such but there would be one returned to us from the plugin adapter which we will use for the upgrade and migration decisions.

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java (hunk 12 above)
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:943]: I was just trying to add some code coverage tests but did not get to complete it yet. The actual code to exercise the api isn't here yet. Let me see if I can add something to perhaps unit test the calls instead.

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java (hunk 11 above)
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:1030]: will work on a better testcase.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+ */
+public class LRRollingUpgradeHandler {
+
+    private volatile boolean isLRReadyForV2 = false;
+    ILogReplicationVersionAdapter versionAdapter;
+    public LRRollingUpgradeHandler(ILogReplicationVersionAdapter versionAdapter) {
+        this.versionAdapter = versionAdapter;
+    }
+
+    public boolean isLRUpgradeInProgress(TxnContext txnContext) {
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:44]: This method does more than just check if upgrade is in progress.  If clusterUpgrade is not in progress, it migrates data in addition to the checks.  I think it will be better to separate the two operations.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+ * ┌───────────▼────────────┐                         ┌──────▼────────────────┐
+ * │ isLRReadyForV2 = false │                         │ isLRReadyForV2 = true │
+ * └───────────┬────────────┘                         └───────┬───────────────┘
+ *             │            ┌──────────────────────┐          │
+ *             └───────────►│ start Log Replicator │◄─────────┘
+ *                          └──────────────────────┘
+ */
+public class LRRollingUpgradeHandler {
+
+    private volatile boolean isLRReadyForV2 = false;
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:38]: would it be better to store this information persistently so that the check need not be performed every time LR starts?  We can do it in another PR...

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/ILogReplicationVersionAdapter.java [hunk 17]
```diff
@@ -12,4 +14,13 @@ public interface ILogReplicationVersionAdapter {
...
     String getVersion();
+
+    /**
+     * @param txnContext - transaction context where we are checking if we need
+     *                   to behave in a backward compatible manner.
+     * @return true if the cluster is undergoing rolling upgrades with
+     * some nodes running an older code base.
+     * false if all the nodes in the cluster are at the same version as me.
+     */
+    boolean isRollingUpgradeInProgress(TxnContext txnContext);
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/ILogReplicationVersionAdapter.java:25]: I think that the method to be exposed here should be getClusterVersion().

isRollingUpgradeInProgress() is our internal method which checks the following:
1) all nodes in the cluster are on the same version?  (depends on Chris's change to add version info per node)
2) node version == clusterVersion (returned from the plugin).

All places in LR invoke isRollingUpgradeInProgress() to decide whether to use the old or new api's.

Additionally, on startup, LR starts a periodic thread which calls isRollingUpgradeInProgress().  If not, and if the atomic boolean is not set, it clears all old data and updates the atomic boolean.  Also, stops this executor service as migration/clear of old data is already done.

Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/ILogReplicationVersionAdapter.java]: Done, please checkout latest commit. thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 16 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java]: Kind of done in the re-work. thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 15 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:44]: Resolved as we discussed this offline. thanks

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java (hunk 12 above)
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:953]: Fixed the test case to validate the new code added.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Rolling upgrade handling means cluster must function in a mode where not all
+ * nodes are running the same codebase.
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once all nodes have upgraded it may atomically migrate data, clean up old code and switch
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which is intended to be used as follows
+ * on LR startup and when upgrade is complete.
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:15]: this will be invoked on every access to the new tables/schema, right?  Not just on startup.  Can you update the description?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        this.versionAdapter = versionAdapter;
+    }
+
+    public boolean isLRUpgradeInProgress(TxnContext txnContext) {
+        if (isClusterAllAtV2) {
+            return false;
+        }
+        String nodeVersion = versionAdapter.getNodeVersion();
+        String pinnedClusterVersion = versionAdapter.getPinnedClusterVersion(txnContext);
+        boolean isClusterUpgradeInProgress = !nodeVersion.equals(pinnedClusterVersion);
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:50]: This check may be sufficient in this version, but it will not work in subsequent versions.

Think about V3..
node 1 version = v2
node 2 version = v2
node 3 version = v3
cluster version = v2.

In this case, upgrade has not yet completed and we must not migrate the data

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultLogReplicationConfigAdapter.java [hunk 20]
```diff
@@ -1,12 +1,28 @@
...
+import org.corfudb.runtime.collections.TxnContext;
+
 /**
  * Default testing implementation of a Log Replication Config Provider
  */
 public class DefaultLogReplicationConfigAdapter implements ILogReplicationVersionAdapter {
 
+    private static final String latestVer = "LATEST-VERSION";
+
+    public DefaultLogReplicationConfigAdapter(CorfuRuntime runtime) {
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultLogReplicationConfigAdapter.java:13]: since we are passing the txn context, is the runtime required?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java [hunk 21]
```diff
@@ -12,33 +16,155 @@ public abstract class DefaultAdapterForUpgrade implements ILogReplicationVersion
...
-                corfuStore.openTable(NAMESPACE, VERSION_TEST_TABLE,
-                        LogReplicationStreams.VersionString.class,
-                        LogReplicationStreams.Version.class, CommonTypes.Uuid.class,
-                        TableOptions.builder().build());
-            } catch (Exception e) {
-                // Just for wrap this up
-            }
+    public DefaultAdapterForUpgrade(CorfuRuntime runtime) {
+        this.corfuStore = new CorfuStore(runtime);
+        try {
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:34]: nit: can we reuse openVersionTable() here?

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java [hunk 22]
```diff
@@ -937,4 +939,46 @@ private void verifyVersion(CorfuStore corfuStore, String expectedVersion, boolea
...
+     * 4. verify that data migration and rolling upgrade end is detected
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testLocalClusterRollingUpgrade() throws Exception {
+        log.info(">> Setup replication for testing during rolling upgrade of active cluster");
+        setupActiveAndStandbyCorfu();
+
+        DefaultAdapterForUpgradeActive.startRollingUpgrade(corfuStoreActive);
```
Comment (pankti-m) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:957]: If we instantiate the default adapter before starting rolling upgrade, do we need to make start and endRollingUpgrade static?

Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:957]: good point - no they need not be - let me make them non-static. thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 18 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:15]: Yes, that is right. I will update the comment. thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java (hunk 19 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:50]: but it is true for v2 right? it means the cluster is already at v2 same as what the current node is.
Also data migration will not be invoked in this case because when LR starts up , the v2 nodes will discover that the pinned version matches their own version and so will automatically bypass any checking right?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultLogReplicationConfigAdapter.java (hunk 20 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultLogReplicationConfigAdapter.java:13]: yes the runtime is needed to instantiate the corfuStore instance which will be used to openTable()s initially. If we do this we can avoid attempting to open tables on access.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java (hunk 21 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:34]: Thanks, done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+ *                     ┌───────────────▼───────────┐
+ *                   ┌─┤                           ├─┐
+ *         Yes ┌─────┴─┴┐  isRollingUpgradeOn() ? ┌┴─┴──────┐ No
+ *             │        └─────────────────────────┘         │
+ *             │                                            │
+ * ┌───────────▼──────────────┐                         ┌──────▼──────────────────┐
+ * │ isClusterAllAtV2 = false │                         │ isClusterAllAtV2 = true │
+ * └───────────┬──────────────┘                         └───────┬─────────────────┘
+ *             │            ┌──────────────────────┐            │
+ *             └───────────►│ start Log Replicator │◄───────────┘
```
[the remark this answers is not in the document]
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:36]: I did not add it in this PR because we don't have any migration code yet, let me see if the startup code can still be modified to match above diagram.

[the remark this answers is not in the document]
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:37]: Done, added the check to LR startup.

**Review by pankti-m — APPROVED** [review]

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java [hunk 24]
```diff
@@ -937,4 +939,46 @@ private void verifyVersion(CorfuStore corfuStore, String expectedVersion, boolea
...
+            Assert.assertTrue(rollingUpgradeHandler.isLRUpgradeInProgress(txnContext));
+        }
+
+        defaultAdapterForUpgradeActive.endRollingUpgrade(corfuStoreActive);
+
+        try (TxnContext txnContext = corfuStoreActive.txn(DefaultAdapterForUpgradeActive.NAMESPACE)) {
+            Assert.assertFalse(rollingUpgradeHandler.isLRUpgradeInProgress(txnContext));
+        }
+
+        if (activeCorfu != null) {
```
Comment (annym) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:971]: Two minor comments:
(1) I know this is not part of your PR but seems reasonable to have  as a cleanup a cleanup method with the 'After' annotation as this should run for every test. 
(2) Is this properly rebased? I believe Shama had renamed all these to source/sink.

**Review by annym — COMMENTED** [review]
Minor comments

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java [hunk 25]
```diff
@@ -937,4 +939,46 @@ private void verifyVersion(CorfuStore corfuStore, String expectedVersion, boolea
...
+     * 1. Simulate startRollingUpgrade()
+     * 2. verify that rolling upgrade is detected
+     * 3. simulate endRollingUpgrade()
+     * 4. verify that data migration and rolling upgrade end is detected
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testLocalClusterRollingUpgrade() throws Exception {
+        log.info(">> Setup replication for testing during rolling upgrade of active cluster");
```
[the remark this answers is not in the document]
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:954]: Shama's source/sink change is in the `log_replication_v2` branch once it merges to master we would fix up all the usages including this one.

[the remark this answers is not in the document]
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:954]: The rolling upgrade handling is independent of source or sink so I only added one of the two and since they extend the base case which has the implementation, it still gives code coverage.

File: test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java (hunk 24 above)
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java:971]: (1) Sounds good - i think in the follow up PR where we add the real code in the `migrateData()` paths we will need a proper cleanup or at least a check
(2) Source sink is pushed to `log_replication_v2` branch

**Review by shama358 — APPROVED** [review]

**Review by annym — APPROVED** [review]

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java [hunk 26]
```diff
@@ -12,33 +16,148 @@ public abstract class DefaultAdapterForUpgrade implements ILogReplicationVersion
...
+        pinnedVersionString =
+                ((LogReplicationStreams.Version)txn.getRecord(VERSION_TEST_TABLE,
+                        versionStringKey).getPayload()).getVersion();
+    }
+
+    @Override
+    public String getPinnedClusterVersion(TxnContext txnContext) {
+        getPinnedVersionString(txnContext);
+        return pinnedVersionString;
+    }
```
Comment (xcchang) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:75]: why we create a separate overloaded method here? seems like the pinnedVersionString could be directly created here?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+ *                   ┌─┤  isRolling UpgradeON?    ├─┐
+ *         Yes ┌─────┴─┴┐  (migrate if done)     ┌┴─┴─────────────┐ No
+ *             │        └────────────────────────┘                │
+ *             │                                                  │
+ * ┌───────────▼────────────────────┐                     ┌───────▼───────────────────────┐
+ * │ Write in both old & new format │                     │ write data in new format only │
+ * └───────────┬────────────────────┘                     └───────┬───────────────────────┘
+ *             │            ┌──────────────────────┐              │
+ *             └───────────►│ end transaction      │◄─────────────┘
+ *                          └──────────────────────┘
```
Comment (xcchang) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java:38]: Did we agree this diagram will be removed?

**Review by xcchang — COMMENTED** [review]
LGTM, just a couple of minor comments

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java (hunk 26 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultAdapterForUpgrade.java:75]: the `getPinnedClusterVersion` is coming from the interface now.
